README.md
<div align="center">
  
  ![Logo](./assets/logo.png)
  
  # MSN Weather API
  A simple MSN Weather API wrapper with built-in TypeScript support.
  
  [![Travis CI](https://img.shields.io/travis/com/wgumenyuk/msn-weather?label=Build&style=flat-square)](https://travis-ci.com/github/wgumenyuk/msn-weather)
  [![Codecov](https://img.shields.io/codecov/c/github/wgumenyuk/msn-weather?label=Coverage&logo=codecov&style=flat-square)](https://codecov.io/gh/wgumenyuk/msn-weather)
  [![Code Climate](https://img.shields.io/codeclimate/maintainability/wgumenyuk/msn-weather?label=Maintainability&logo=Code%20Climate&style=flat-square)](https://codeclimate.com/github/wgumenyuk/msn-weather)
  [![NPM downloads](https://img.shields.io/npm/dt/msn-weather?label=Downloads&style=flat-square)](https://www.npmjs.com/package/msn-weather)
  [![Size](https://img.shields.io/bundlephobia/minzip/msn-weather?label=Size&style=flat-square)](https://github.com/wgumenyuk/msn-weather)
  [![License](https://img.shields.io/github/license/wgumenyuk/msn-weather?label=License&style=flat-square)](./LICENSE)
  
  [![NPM install info](https://nodei.co/npm/msn-weather.png?downloads=true&stars=true)](https://www.npmjs.com/package/msn-weather)
</div>

## Table of contents
- [About](#about)
- [Installation](#installation)
- [Usage](#usage)
  - [Usage with CommonJS](#usage-with-commonjs)
  - [Retrieved data format](#retrieved-data-format)
  - [API documentation](#api-documentation)
- [Other languages](#other-languages)
- [Resources](#resources)
- [Acknowledgements](#acknowledgements)
  - [Contributors](#contributors)
- [License](#license)

## About
`msn-weather` is a powerful [Node.js](https://nodejs.org) library that allows you to easily retrieve weather data for any location in the world.
As the name suggests, this wrapper uses the MSN weather API behind the scenes.

- Simple and easy-to-use API
- Built-in TypeScript definitions
- Only one dependency
- Clear weather data format
- Performant

## Installation
Install this package using NPM:

```sh-session
npm install msn-weather --save
```

## Usage
```ts
import weather from "msn-weather";

const data = await weather.search({
  location: "Munich, DE",
  language: "en",
  degreeType: "C"
});
```

### Usage with CommonJS
To use this library with CommonJS, use this approach:
```js
const weather = require("msn-weather").default;
```

### Retrieved data format
You will receive a JavaScript object looking like this:

<details>
  <summary>Show response</summary>
  
  ```js
  {
    current: {
      date: "2021-06-09",
      day: "Wednesday",
      temperature: "23°C",
      sky: {
        code: "partly_sunny",      
        text: "Partly Sunny"        
      },
      observation: {
        time: "12:00:00",
        point: "Munich, BY, Germany"
      },
      feelsLike: "22°C",
      humidity: "58%",
      wind: {
        display: "7 km/h North",    
        speed: "7 km/h"
      }
    },
    forecasts: [
      {
        date: "2021-06-09",
        day: "Wednesday",
        temperature: {
          low: "14°C",
          high: "24°C"
        },
        sky: {
          code: "partly_sunny",
          text: "Partly Sunny"
        },
        precip: "90%"
      },
      {
        date: "2021-06-10",
        day: "Thursday",
        temperature: {
          low: "13°C",
          high: "22°C"
        },
        sky: {
          code: "light_rain",
          text: "Light Rain"
        },
        precip: "100%"
      },
      {
        date: "2021-06-11",
        day: "Friday",
        temperature: {
          low: "15°C",
          high: "23°C"
        },
        sky: {
          code: "light_rain",
          text: "Light Rain"
        },
        precip: "100%"
      },
      {
        date: "2021-06-12",
        day: "Saturday",
        temperature: {
          low: "15°C",
          high: "24°C"
        },
        sky: {
          code: "light_rain",
          text: "Light Rain"
        },
        precip: "100%"
      }
    ]
  }
  ```
</details>

### API documentation
#### `weather.search(options)`
Retrieves weather data for a given location. Returns a promise with weather data (see [retrieved data format](#retrieved-data-format)).

| Parameter | Type                  | Optional | Default | Description             |
|-----------|-----------------------|----------|---------|-------------------------|
| `options` | [`Options`](#options) | ❌       | None    | Options for the search. |

#### `Options`
Options for the search.
| Parameter | Type | Optional | Default | Description |
|-|-|-|-|-|
| `location` | String | ❌ | None | Location for the weather data. The location will be encoded automatically using [`encodeURIComponent`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent). |
| `language` | String | ✔  | `en`  | Language in which weather text will be returned. The value must be a [ISO 639.1:2002 language code](https://en.wikipedia.org/wiki/ISO_639-1). |
| `degreeType` | String | ✔ | `C` | Degree type for temperature values. Either Celsius (`C`) or Fahrenheit (`F`). |

## Resources
- [Changelog](./CHANGELOG.md)
- [NPM](https://www.npmjs.com/package/msn-weather)
- [GitHub](https://github.com/wgumenyuk/msn-weather)

## Acknowledgements
- [Logo](https://twemoji.twitter.com) by Twitter Twemoji (licensed under [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/))

### Contributors
A special thanks goes out to these contributors:

- Khang ([khang-nd](https://github.com/khang-nd)) - Contributing a number of times

## License
This project is licensed under [MIT](./LICENSE).

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. This project adheres to [Semantic Versioning](https://semver.org/).

## Unreleased
- Changed `textID` 28 from `partly_sunny` to `mostly_cloudy`

## 2.1.3 (June 04, 2022)
### Fixed
- Faulty index access giving back first character only

### Changed
- Changed `textID` 11 from `light_rain` to `rain_shower`

## 2.1.2 (August 01, 2021)
### Added
- Added a contributors anchor
- Added a [`encodeURIComponent`](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent) explanation to API documentation

### Changed
- Replaced version badge with size badge

## 2.1.1 (July 20, 2021)
### Changed
- Switched from `xml2js` to `fast-xml-parser`: this leads to a much smaller bundle size (see [xml2js](https://bundlephobia.com/package/xml2js@0.4.23) vs. [fast-xml-parser](https://bundlephobia.com/package/fast-xml-parser@3.19.0))
- Renamed `utils` directory to `core`
- Renamed `data` directory to `expected`
- Improved tests

## 2.0.1 (July 16, 2021)
### Fixed
- Fixed bad JSDoc for `request` function

## 2.0.0 (July 02, 2021)
### Added
- Added anchor link to retreived data format to describe the returned data better
- Added contributor acknowledgements to README

### Changed
- Changed HTTP to HTTPS ([#2](https://github.com/wgumenyuk/msn-weather/issues/2))

### Fixed
- Broken anchor link in README

## 1.0.2 (June 20, 2021)
### Fixed
- Fixed wrong API documentation ([#1](https://github.com/wgumenyuk/msn-weather/pull/1))
- Fixed broken link in README

### Changed
- Converted language code to lowecase in README

## 1.0.1 (June 10, 2021)
### Added
- Added acknowledgements to README

### Fixed
- Fixed some README typos

## 1.0.0 (June 10, 2021)
### Added
- `search()` method to retrieve weather data 
- Detailed README


tests/api.test.ts
import fs from "fs";
import path from "path";
import nock from "nock";
import weather from "../src";
import request from "../src/core/request";

const msnMock = nock("https://weather.service.msn.com/find.aspx");

const mockDataPath = path.join(__dirname, "./expected/mock.xml");
const mockData = fs.readFileSync(mockDataPath, "utf-8");

const url =
    "https://weather.service.msn.com/find.aspx?src=msn&" +
    "weadegreetype=C&" +
    "culture=en&" +
    `weasearchstr=${encodeURIComponent("München, DE")}`;

describe("msn-weather", () => {
    describe("msn-weather#search", () => {
        it("should be a function", (done: jest.DoneCallback) => {
            expect(typeof weather.search).toBe("function");
            done();
        });

        it("should not accept invalid options", async () => {            
            await expect(weather.search(null))
                .rejects
                .toThrow("Invalid options were specified");

            await expect(weather.search({ location: "" }))
                .rejects
                .toThrow("No location was given");

            await expect(weather.search({ location: "München, DE", language: "ABC" }))
                .rejects
                .toThrow("Invalid language code 'ABC', make sure it adheres to the ISO 639.1:2002 standard");

            // @ts-ignore
            await expect(weather.search({ location: "München, DE", degreeType: "K" }))
                .rejects
                .toThrow("Invalid degree type 'K'");
        });

        it("should fail when receiving a non-200 status code", async () => {
            msnMock
                .get("")
                .query({
                    src: "msn",
                    weadegreetype: "C",
                    culture: "en",
                    weasearchstr: "München, DE"
                })
                .reply(404);

            await expect(request(url))
                .rejects
                .toThrow("Request failed with status 404");
        });

        it("should fail when timing out after 5 seconds", async () => {
            msnMock
                .get("")
                .query({
                    src: "msn",
                    weadegreetype: "C",
                    culture: "en",
                    weasearchstr: "München, DE"
                })
                .delayConnection(5100)
                .reply(500);

            await expect(request(url))
                .rejects
                .toThrow("Request timed out after 5s");

            nock.abortPendingRequests();
        });

        it("should fail when a connection error occurs", async () => {
            msnMock
                .get("")
                .query({
                    src: "msn",
                    weadegreetype: "C",
                    culture: "en",
                    weasearchstr: "München, DE"
                })
                .replyWithError("ECONNRESET");

            await expect(request(url))
                .rejects
                .toThrow("ECONNRESET");
        });

        it("should fail if the response body can't be parsed", async () => {    
            msnMock
                .get("")
                .query({
                    src: "msn",
                    weadegreetype: "C",
                    culture: "en",
                    weasearchstr: "München, DE"
                })
                .reply(200);
            
            const options = {
                location: "München, DE"
            };

            await expect(weather.search(options))
                .rejects
                .toThrow("Bad response: Failed to parse response body");
        });

        it("should fail if the response body is incomplete", async () => {    
            msnMock
                .get("")
                .query({
                    src: "msn",
                    weadegreetype: "C",
                    culture: "en",
                    weasearchstr: "München, DE"
                })
                .reply(200,
                    `<weatherdata
                        xmlns:xsd="http://www.w3.org/2001/XMLSchema"
                        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
                        <weather>
                            <current/>
                        </weather>
                    </weatherdata>`
                );
            
            const options = {
                location: "München, DE"
            };

            await expect(weather.search(options))
                .rejects
                .toThrow("Bad response: Failed to receive weather data");
        });

        it("should return correct weather data for a location", async () => {        
            msnMock
                .get("")
                .query({
                    src: "msn",
                    weadegreetype: "C",
                    culture: "en",
                    weasearchstr: "München, DE"
                })
                .reply(200, mockData);
            
            const data = await weather.search({
                location: "München, DE"
            });

            expect(data).toHaveProperty("current");
            expect(data).toHaveProperty("forecasts");

            expect(data.current).toHaveProperty("date");
            expect(data.current).toHaveProperty("day");
            expect(data.current).toHaveProperty("temperature");
            expect(data.current).toHaveProperty("sky");
            expect(data.current.sky).toHaveProperty("code");
            expect(data.current.sky).toHaveProperty("text");
            expect(data.current).toHaveProperty("observation");
            expect(data.current.observation).toHaveProperty("time");
            expect(data.current.observation).toHaveProperty("point");
            expect(data.current).toHaveProperty("feelsLike");
            expect(data.current).toHaveProperty("humidity");
            expect(data.current).toHaveProperty("wind");
            expect(data.current.wind).toHaveProperty("display");
            expect(data.current.wind).toHaveProperty("speed");

            expect(Array.isArray(data.forecasts)).toBeTruthy();
            expect(data.forecasts[0]).toHaveProperty("date");
            expect(data.forecasts[0]).toHaveProperty("day");
            expect(data.forecasts[0]).toHaveProperty("temperature");
            expect(data.forecasts[0].temperature).toHaveProperty("low");
            expect(data.forecasts[0].temperature).toHaveProperty("high");
            expect(data.forecasts[0]).toHaveProperty("sky");
            expect(data.forecasts[0].sky).toHaveProperty("code");
            expect(data.forecasts[0].sky).toHaveProperty("text");
            expect(data.forecasts[0]).toHaveProperty("precip");
        });
    });
});


src/index.ts
import x2js from "fast-xml-parser";
import { isDegreeType, isLanguageCode } from "./core/validators";
import request from "./core/request";
import textIDs from "./core/textIDs";

// #region Types and interfaces
type Degree = "C" | "F";

interface Options {
    location: string;
    language?: string;
    degreeType?: Degree;
}

interface Forecast {
    date: string;
    day: string;
    temperature: {
        low: string;
        high: string;
    };
    sky: {
        code: string;
        text: string;
    };
    precip: string;
}

interface Current {
    date: string;
    day: string;
    temperature: string;
    sky: {
        code: string;
        text: string;
    };
    observation: {
        time: string;
        point: string;
    };
    feelsLike: string;
    humidity: string;
    wind: {
        display: string;
        speed: string;
    };
}

interface Weather {
    current: Current;
    forecasts: Forecast[];
}
// #endregion

/**
    Retreives weather data for a given location.
    @param options              Options.
    @param options.location     The location you're looking for.
    @param options.language     The language in which text will be returned.
    @param options.degreeType   Degree type for temperature values.
    @returns Weather data.
*/
async function search(options: Options): Promise<Weather> {
    if(!options || typeof options !== "object") {
        throw new Error("Invalid options were specified");
    }
    
    if(!options.location) {
        throw new Error("No location was given");
    }

    const language = options.language || "en";
    const degreeType = options.degreeType || "C";

    if(!isLanguageCode(language)) {
        throw new Error(`
            Invalid language code '${language}', make sure it adheres to the ISO 639.1:2002 standard
        `);
    }

    if(!isDegreeType(degreeType)) {
        throw new Error(`Invalid degree type '${degreeType}'`);
    }

    const url =
        "https://weather.service.msn.com/find.aspx?src=msn&" +
        `weadegreetype=${degreeType}&` +
        `culture=${language}&` +
        `weasearchstr=${encodeURIComponent(options.location)}`;

    const response = await request(url);

    const json = x2js.parse(response, {
        attributeNamePrefix: "",
        ignoreAttributes: false,
        ignoreNameSpace: true,
        trimValues: true
    });

    if(!json || !json.weatherdata || !json.weatherdata.weather) {
        throw new Error("Bad response: Failed to parse response body");
    }

    const data = json.weatherdata.weather[0];

    if(!data) {
        throw new Error("Bad response: Failed to receive weather data");
    }

    const current = data.current;
    const forecasts = [];

    for(let i = 1; i < data.forecast.length; i++) { 
        const forecast = data.forecast[i];

        forecasts.push({
            date: forecast.date,
            day: forecast.day,
            temperature: {
                low: forecast.low[0] + `°${degreeType}`,
                high: forecast.high[0] + `°${degreeType}`,
            },
            sky: {
                code: textIDs[forecast.skycodeday[0]],
                text: forecast.skytextday[0]
            },
            precip: forecast.precip[0] + "%"
        });
    }

    const weather: Weather = {
        current: {
            date: current.date,
            day: current.day,
            temperature: current.temperature + `°${degreeType}`,
            sky: {
                code: textIDs[current.skycode],
                text: current.skytext
            },
            observation: {
                time: current.observationtime,
                point: current.observationpoint
            },
            feelsLike: current.feelslike + `°${degreeType}`,
            humidity: current.humidity + "%",
            wind: {
                display: current.winddisplay,
                speed: current.windspeed
            }
        },
        forecasts
    };

    return weather;
}

export default {
    search
};

src/utils/request.ts
import https from "https";

/**
    Sends a request using the native `https` library.
    @param url The target URL.
    @returns HTTPS response body.
*/
function request(url: string): Promise<string> {
    return new Promise((resolve, reject) => {
        const req = https.get(url, (res) => {
            let data = "";

            if(res.statusCode !== 200) {
                reject(new Error(`Request failed with status ${res.statusCode}`));
            }
    
            res.on("data", (chunk: string) => {
                data += chunk;
            });
    
            res.on("end", () => {
                resolve(data);
            });
        });

        req.setTimeout(5000);

        req.on("timeout", () => {
            reject(new Error("Request timed out after 5s"));
        });

        req.on("error", (error) => {
            reject(error);
        });
    });
}

export default request;

src/utils/convertID.ts
function convertID(code: string): string {
    switch(code) {
        case "0":
        case "1":
        case "2":
        case "3":
        case "4":
        case "17":
        case "35": {
            return "thunderstorm";
        }
        
        case "5": {
            return "rain_snow_mix";
        }

        case "6": {
            return "sleet_snow_mix";
        }

        case "7": {
            return "rain_snow_sleet_mix";
        }

        case "8":
        case "9": {
            return "icy";
        }

        case "10": {
            return "rain_sleet_mix";
        }

        case "11": {
            return "light_rain";
        }

        case "12": {
            return "rain";
        }

        case "13": {
            return "light_snow";
        }

        case "14":
        case "16":
        case "42":
        case "43": {
            return "snow";
        }

        case "15": {
            return "blizzard";
        }

        case "18":
        case "40": {
            return "rain_showers";
        }

        case "19": {
            return "dust";
        }

        case "20": {
            return "foggy";
        }

        case "21": {
            return "haze";
        }

        case "22": {
            return "smoke";
        }

        case "23":
        case "24": {
            return "windy";
        }

        case "25": {
            return "frigid";
        }

        case "26": {
            return "cloudy";
        }

        case "27":
        case "29":
        case "33": {
            return "mostly_cloudy_night";
        }

        case "28":
        case "30":
        case "34": {
            return "mostly_cloudy";
        }

        case "31": {
            return "clear_night";
        }

        case "32": {
            return "clear";
        }

        case "36": {
            return "hot";
        }

        case "37":
        case "38": {
            return "scattered_thunderstorms";
        }

        case "39": {
            return "scattered_rain_showers";
        }

        case "41": {
            return "scattered_snow_showers";
        }

        case "45": {
            return "scattered_rain_showers_night";
        }

        case "46": {
            return "scattered_snow_showers_night";
        }

        case "47": {
            return "scattered_thunderstorms_night";
        }

        default: {
            return "N/A";
        }
    }
}

export default convertID;

src/core/textIDs.ts
/**
    A list of custom text IDs for MSN's number codes.
*/
const textIDs: { [key: string]: string } = {
    "0": "thunderstorm",
    "1": "thunderstorm",
    "2": "thunderstorm",
    "3": "thunderstorm",
    "4": "thunderstorm",
    "17": "thunderstorm",
    "35": "thunderstorm",
    "5": "rain_snow_mix",
    "6": "sleet_snow_mix",
    "7": "rain_snow_sleet_mix",
    "8": "light_rain",
    "9": "light_rain",
    "10": "rain_sleet_mix",
    "11": "rain_shower",
    "12": "rain",
    "13": "light_snow",
    "14": "snow",
    "16": "snow",
    "42": "snow",
    "43": "snow",
    "15": "blizzard",
    "18": "rain_showers",
    "40": "rain_showers",
    "19": "dust",
    "20": "foggy",
    "21": "haze",
    "22": "smoke",
    "23": "windy",
    "24": "windy",
    "25": "frigid",
    "26": "cloudy",
    "27": "mostly_cloudy_night",
    "29": "mostly_cloudy_night",
    "33": "mostly_cloudy_night",
    "28": "mostly_cloudy",
    "30": "partly_sunny",
    "34": "partly_sunny",
    "31": "clear_night",
    "32": "clear",
    "36": "hot",
    "37": "scattered_thunderstorms",
    "38": "scattered_thunderstorms",
    "39": "scattered_rain_showers",
    "41": "scattered_snow_showers",
    "45": "scattered_rain_showers_night",
    "46": "scattered_snow_showers_night",
    "47": "scattered_thunderstorms_night"
};

export default textIDs;